keps/sig-storage/2644-honor-pv-reclaim-policy/README.md

```diff
@@ -0,0 +1,421 @@
+		return false
+	} else if volume.ObjectMeta.DeletionTimestamp != nil { //<========== DeletionTimestamp is set on the PV since it's deleted first.
+		return false
+	}
+    // Remaining code below skipped for brevity...
+    // Please refer sig-storage-lib-external-provisioner/controller/controller.go for the full code.
+```
+
+The main approach in fixing the issue involves introducing new finalizer `kubernetes.io/pv-deletion-protection` on the
+Persistent Volume when it is observed that it has `DeletionTimestamp` set but still associated with a valid claim.
+Adding the finalizer prevents the PV to be removed from the API server.
+
+When the PVC is deleted, the PV is moved into `Released` state and following checks are made:
+
+1. Plugin:
+If the volume has the finalizer `kubernetes.io/pv-deletion-protection`, then, `DeletionTimestamp` checks can be ignored.
+
+3. CSI driver
+
+If when processing the PV update it is observed that it has `kubernetes.io/pv-deletion-protection` finalizer and
+`DeletionTimestamp` set, then the volume deletion request is forwarded to the driver(provided other pre-defined 
+conditions are met).
+
+
+
+### Risks and Mitigations
+
+The primary risk is volume deletion to a user that expect the current behavior, i.e, they do not expect the volume to be
+deleted from the underlying storage infrastructure when for a bound PV-PVC pair, the PV is deleted followed by deleting
+the PVC. This can be mitigated by initially introducing the fix behind a feature gate, followed by explicitly deprecating
+the current behavior.
+
+## Design Details
+
+When processing the PV update in `kube-controller-manager`, if it's observed that the PV is still bound to a valid claim,
+and also has `DeletionTimestamp` set, then a finalizer is added to the PV.
+
+```go
+// PvDeletionProtectionFinalizer finalizer is added to the PV when it is observed that the PV
+// has a deletion timestamp but still associated with a valid claim.
+PvDeletionProtectionFinalizer = "kubernetes.io/pv-deletion-protection"
+```
+
+```go
+// Function: syncVolume
+// Code above skipped for brevity...
+if !found {
+			klog.V(4).Infof("synchronizing PersistentVolume[%s]: claim %s not found", volume.Name, claimrefToClaimKey(volume.Spec.ClaimRef))
+			// Fall through with claim = nil
+		} else {
+			var ok bool
+			claim, ok = obj.(*v1.PersistentVolumeClaim)
+			if !ok {
+				return fmt.Errorf("Cannot convert object from volume cache to volume %q!?: %#v", claim.Spec.VolumeName, obj)
+			}
+			klog.V(4).Infof("synchronizing PersistentVolume[%s]: claim %s found: %s", volume.Name, claimrefToClaimKey(volume.Spec.ClaimRef), getClaimStatusForLogging(claim))
+			// Check if the PV has a deletion timestamp, if so, annotate the PV to indicate that the PV can be deleted
+			// from the underlying storage.
+			if protectionutil.NeedToAddFinalizer(volume, pvutil.PvDeletionProtectionFinalizer) {
```

Comment:
When the PV protection controller and PV controller (this code) are aligned badly, PV may be already deleted at this point.

1. User creates PVC + gets PV provisioned.
2. User deletes PV. PV protection finalizer keeps the PV alive.
3. User deleted PVC.
4. PV protection controller sees that there is no PV and deletes its finalizer. The API server deletes the PV at this point
5. PV controller (this code) gets the updated PV late and at this point the PV does not exist any longer. If the controller is restarted, all knowledge about the PV is lost.

A fix is relatively simple: The deletion protection finalizer must be added earlier, right when the PV is provisioned or when it's seen (by PV controller?) with `ReclaimPolicy: Delete` and without the finalizer and without DeletionTimestamp. The finalizer should be also removed when user changes `ReclaimPolicy` to `Retain` or ` Recycle`.

Maybe it's easier to re-frame this enhancement from "detect that PV was deleted before PVC and make sure the volume is deleted" to "make sure the PV is always deleted". There is a subtle difference here - you don't need to detect that a PV was deleted first. Just put the finalizer to the PV the first time the controller sees it and remove it only after the volume is deleted.